x86 : Memory domain implementation for x86

AdithyaBaglody opened the pull request:
This PR shows the implementation of memory domain for x86.

P.S.
In the PR #1516 there were some build related issues which were not caught. This PR fixes that and also reverses the reverts that happened for the PR.

Delta from PR #1516 
1. Added prototype for arm_core_mpu_remove_mem_partition in include/arch/arm/cortex_m/mpu/arm_core_mpu_dev.h

AdithyaBaglody committed:
kernel: arch interface for memory domain

Additional arch specific interfaces to handle memory domain
destroy and single partition removal.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
ARM: MPU: Arch specific memory domain APIs

Added architecture specific support for memory domain destroy
and remove partition for arm and nxp. An optimized version of
remove partition was also added.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
x86: MMU: Memory domain implementation for x86

Added support for memory domain implementation.

GH-3852

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
x86: MMU: Configure page tables entries for memory domain in swap.

During swap the required page tables are configured. The outgoing
thread's memory domain pages are reset and the incoming thread's
memory domain is loaded. The pages are configured if userspace
is enabled and if memory domain has been initialized before
calling swap.

GH-3852

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
kernel: Arch specific memory domain APIs added

Added arch specific calls to handle memory domain destroy
and removal of partition.

GH-3852

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
samples: MPU: updated the memory domain test for x86.

GH-3852

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

andrewboie commented:
Needs a rebase.

Some weird test case errors too:

```
***** CPU Page Fault (error code 0x00000000)
Supervisor thread read address 0xaaaaaaaa
PDE: 0x000 Non-present Read-only Supervisor
PTE: 0x000 Non-present Read-only Supervisor
Current thread ID = 0x0041b7a4
Faulting segment:address = 0x0008:0x00008606
eax: 0xaaaaaaaa, ebx: 0x00401400, ecx: 0x00419000, edx: 0x00000001
esi: 0x00000001, edi: 0x00401460, ebp: 0x0041b764, esp: 0x0041b754
eflags: 0x86
Fatal fault in essential thread! Spinning...
Terminate emulator due to fatal kernel error
```
Not sure what's going on here, but 0xAAAAAAAA is what we initialize stack memory regions to if CONFIG_INIT_STACKS=y

tyrantlizard77 reviewed (changes requested):
rebase please, also need CI passing

AdithyaBaglody commented:
Delta:
1. When added the x86 pae mode on top these commits there was a overflow happening with the attributes of the partition. Changed it to support different size of the attributes.
2. due to a problem in the arm build system. I am waiting on the PR #4542 to get merged.

PS shippable will complain until that is merged.

nashif commented:
recheck

tyrantlizard77 reviewed (dismissed):
You need to update sample.yaml for this test case it doesn't get build for x86.
You could remove the platform_whitelist and instead put "filter: CONFIG_ARCH_HAS_USERSPACE"

andrewboie commented:
sample.yaml taken care of in #4678 
This needs to be rebased though

andrewboie commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -52,4 +52,37 @@ int _arch_mem_domain_max_partitions_get(void)
 }
+
+/**
+ * @brief Reset MPU region for a single memory partition
```
you don't need to put documentation here.
document this in the arch-independent header

andrewboie commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -52,4 +52,37 @@ int _arch_mem_domain_max_partitions_get(void)
+}
+
+/**
+ * @brief Destroy MPU regions for the mem domain
```
same as above

andrewboie commented on kernel/include/nano_internal.h:
```diff
@@ -95,6 +95,30 @@ static inline unsigned int _Swap(unsigned int key)
+ * @param domain The memory domain structure
+ * @param partition_id The partition that needs to be deleted
+ */
+extern void _arch_mem_domain_remove_partition(struct k_mem_domain *domain,
```
Zephyr coding standard is to have the verb last.
Should be _arch_mem_domain_partition_remove

AdithyaBaglody commented:
Delta:  Handled review comments from Andrew.
Pending: Header file inclusion sequence issue. Shippable will continue to complain

lpereira commented on samples/mpu/mem_domain_apis_test/src/main.c:
```diff
@@ -26,13 +26,23 @@ struct k_thread app_thread_id[3];
 u8_t __aligned(32) app0_buf[32];
 u8_t __aligned(32) app1_buf[32];
+#else
+u8_t __aligned(4096) app0_buf[4096];
```
Can we have a `K_MEM_PARTITION_ALIGNED` (or something of the sort) that each platform specifies with the correct page size?

andrewboie replied:
It's hard to capture this in a completely generic way.
On x86, memory is configured at 4K page granularity, the regions have to be 4K aligned but the size can be any multiple of 4K.
With ARM MPU, a region can be of any power-of-two size starting at 32 bytes, and must be aligned to its size.
Other MPUs may be more permissive.
This sample needs to be rewritten as a test case anyway as it doesn't actually prove that anything works in terms that CI can understand.

andrewboie replied:
Having said this, one cheap way to get rid of the ifdef is just make the size 4096, aligned to 4096, that ought to work everywhere.

andrewboie commented:
Inclusion issues should now be resolved if you can rebase this.

AdithyaBaglody committed:
kernel: init: mem_domain structure is initialized for dummy thread.

For the dummy thread, contents in the mem_domain structure
is insignificant hence setting it to NULL.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

andrewboie commented:
build failing due to documentation issues

> /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/doc/api/kernel_api.rst:232: WARNING: doxygengroup: Cannot find namespace "mem_domain_apis" in doxygen xml output for project "Zephyr" from directory: doxygen/xml

dbkinder commented on include/kernel.h:
```diff
@@ -4180,6 +4180,7 @@ static inline char *K_THREAD_STACK_BUFFER(k_thread_stack_t *sym)
 
 #endif /* _ARCH_DECLARE_STACK */
 
+#if defined(CONFIG_USERSPACE)
```
the doxygen group "mem_domain_apis" will not be defined if CONFIG_USERSPACE is not defined, and there are references to this doxygen group in doc/api/kernel_api.rst  that will throw an error.

andrewboie replied:
IIRC, the way we do this with other conditionally compiled subsystems is to not ifdef the API definitions in the headers, but instead just conditionally compile the implementation, so that the user gets errors at link time if they try to use something that isn't enabled.

AdithyaBaglody replied:
I understand the issue now, I was wondering what was causing this.
@dbkinder why are we doing preprocessing when generating doc using doxygen? is there a specific reason? incidentally we are missing out on a lot of information since a lot the code would depend on a config. Can we do, "ENABLE_PREPROCESSING = y" for the doxygen?. I think that was the control for doxygen, AFAIK.

AdithyaBaglody replied:
@andrewboie I think there might be some issue regarding that solution. I want to hide the k_mem_domain and k_mem_partition under CONFIG_USERSPACE. But the mem_domain api needs these structures to even compile. Thus causing a compilation issue and not a linker issue. I will see if there is a better way.

andrewboie replied:
what is the point of hiding these structs?
just get rid of the ifdef!

dbkinder replied:
ENABLE_PREPROCESSING=YES is the doxygen setting we're using (see doc/zephyr.doxyfile)

dbkinder replied:
ENABLE_PREPROCESSING has been enabled since I've been with the project... Not sure of the impact of removing it but I can check and see how it would impact the documentation.

andrewboie replied:
enabling preprocessing will not help you because CONFIG_USERSPACE is not on by default

@AdithyaBaglody please get rid of the ifdef. We don't care about exposing the structs. We don't conditionally compile any of the k_poll API definitions even though the implementation is selectively compiled with K_POLL

AdithyaBaglody commented:
Delta:
Handled the David's comments on documentation.
